protosanitizer/protosanitizer.go

```diff
@@ -48,67 +48,68 @@ type stripSecrets struct {
 }
 
 func (s *stripSecrets) String() string {
-	// First convert to a generic representation. That's less efficient
-	// than using reflect directly, but easier to work with.
-	var parsed interface{}
-	b, err := json.Marshal(s.msg)
-	if err != nil {
-		return fmt.Sprintf("<<json.Marshal %T: %s>>", s.msg, err)
-	}
+	stripped := s.msg
+
+	// also support scalar types like string, int, etc.
 	msg, ok := s.msg.(proto.Message)
-	if !ok {
-		return string(b)
-	}
-	if err := json.Unmarshal(b, &parsed); err != nil {
-		return fmt.Sprintf("<<json.Unmarshal %T: %s>>", s.msg, err)
+	if ok {
+		stripped = stripMessage(msg.ProtoReflect())
 	}
 
-	// Now remove secrets from the generic representation of the message.
-	s.strip(parsed, msg.ProtoReflect())
-
-	// Re-encoded the stripped representation and return that.
-	b, err = json.Marshal(parsed)
+	b, err := json.Marshal(stripped)
 	if err != nil {
 		return fmt.Sprintf("<<json.Marshal %T: %s>>", s.msg, err)
 	}
 	return string(b)
 }
 
-func (s *stripSecrets) strip(parsed interface{}, msg protoreflect.Message) {
-	// The corresponding map in the parsed JSON representation.
-	parsedFields, ok := parsed.(map[string]interface{})
-	if !ok {
-		// Probably nil.
-		return
+func stripSingleValue(field protoreflect.FieldDescriptor, v protoreflect.Value) any {
+	switch field.Kind() {
+	case protoreflect.MessageKind:
+		return stripMessage(v.Message())
+	case protoreflect.EnumKind:
+		return field.Enum().Values().ByNumber(v.Enum()).Name()
+	default:
+		return v.Interface()
 	}
+}
+
+func stripValue(field protoreflect.FieldDescriptor, v protoreflect.Value) any {
```

Comment:
Can be `stripValue()` and `stripSingleValue()` combined into a single function with a bigger switch?

```go
switch {
case field.IsList():
...

case field.IsMap():
...

case field.Kind() == protoreflect.MessageKind:
...
case field.Kind() == ...

default: return v.Interface()
```

Reply:
It is possible, but less clear, I think. Because for field that `IsList()`, its `Kind()` still returns the kind of list element. So, we will have multiple case statements matched.

My code structure actually looks very like the official encoder:
https://github.com/protocolbuffers/protobuf-go/blob/b98563540c0a4edb38526bcd6e6c97f9fac1f453/encoding/prototext/encode.go#L201-L212